rfcs/2021-07-20-8288-csv-enrichment.md

```diff
@@ -0,0 +1,321 @@
+database connections.
+
+*path*
+
+The path to the csv file, either an absolute path or one relative to the current
+working directory.
+
+*delimiter*
+
+The delimiter used in the csv file to separate fields. Defaults to `","`.
+
+*title_row*
+
+If true, it assumes the first row in the csv file contains column names. If
+false, columns are named according to their numerical index.
+
+Note, this is likely to be where any encryption parameters such as the key are
+specified when it comes time to implementing encryption.
+
+The initial implementation will only be supporting CSV files as a resource. It
+is anticipated that future work will expand the available resources to include
+other file types as well as databases.
+
+Like the services that sources and sinks integrate with, enrichment "tables"
+will likely need administrator approval and setup. Typically data is enriched
+from some sort of shared company "resource" that end users will likely not have
+access to. Admin will likely want to restrict access to resources to approved
+pipelines, especially if sensitive information is contained.
+
+The tables will need to be integrated with `vector validate`, which would
+ensure the resource exists and is correctly formatted.
+
+### Schema
+
+For the CSV Table all columns will be considered to be Strings. Since Tables are
+loaded before VRL compilation it will be possible to ensure that Vrl doesn't
+search on columns that do not exist within the datafile. Searching on a column
+that doesn't exist can prevent Vector from loading.
+
+If (when these features are implemented) the user attempts to reload a Table
+and a column that has been Indexed no longer exists, this should prevent the
+file from being loaded. Vector will continue to use the currently loaded data.
+
+These restrictions apply only to columns that VRL is searching on. If the user
+attempts to use a field that has not been returned from the enrichment, this
+will be possible, but the value will be Null. This is largely due to the way
+that VRL works with paths into Objects.
+
+### Vrl functions
+
+Two remap functions:
+
+#### `find_table_row`
+
+This function will look up a single row within the table dataset. If a single
+row is found that data is returned as an object, otherwise this function will
+error.
+
+A metric will be emitted to indicate the lookup time.
+
+#### `search_table_rows`
+
+This function returns the results as an array of objects. If no row is found an
+empty array is returned. Multiple results just result in multiple rows in the
+array.
+
+A metric will be emitted to indicate the lookup time.
+
+#### Parameters
+
+*table*
+
+The name of the table to lookup. This must point to a table specified in the
+config file eg `tables.csv_file`. Both functions are generic over all Table
+types.
+
+*condition*
+
+`condition` is a single level, key/value object that specifies the fields and
+values to lookup. The fields must all match (AND) for the row to be returned.
```

Comment:
Could we use named args here instead of an object? E.g. `find_table_row(tables.csv_file, license_plate: .license)`. That seems a little more natural and emphasizes that the lookup key (i.e. `license_plate`) is static.

I'd also lean toward starting out with a single key lookup for this. If there's user demand, we can add multiple keys later.

Reply:
I'm cool with starting with a single key but the user that motivated this feature does specifically have a need for composite keys.

For named arguments, we don't currently have support dynamic keys (and hesitate to introduce that as part of this feature). I think we could do `find_table_row(tables.csv_file, { "license_plate": .license } )` though.

Reply:
Ah, good to know. A single composite key lookup is still quite a bit simpler than supporting multiple lookup keys. Basically, the hierarchy of complexity goes:

1. Single scalar key:
```
find_table_row(..., { "license": .license })
```
2. Single composite key:
```
find_table_row(..., { "license": .license, "name": .name })
```
3. Multiple keys:
```
find_table_row(..., { "license": .license })
find_table_row(..., { "name": .name })
```

I'm just suggesting that we only go as high as we need to for the first iteration to be useful.

Reply:
That doesn't fit so well with the way VRL has been designed, so would require a more extensive change to the language. It's certainly doable if it is something you feel strongly about, but if not I would prefer to stick with an object for now.

We already know there is user demand for multiple keys, so do need to make this work.

In terms of indexing, since all lookups are AND, it should be a fairly straightforward operation to  index based on the combined hash of the fields to lookup. This would get more complicated should more complicated searches be required, but we can defer that to a later date should that happen.